### PR TITLE
Safer history restoration

### LINKED ISSUE #2602: Nyxt does not start without history/default.lisp
**Describe the bug**
Nyxt will not start if default history file is not available.

**Precise recipe to reproduce the issue**
Sometimes to try out a new nyxt version I delete everything user specific:
- ~/.config/nyxt
- ~/.cache/nyxt
- ~/.local/share/nyxt

With recent git versions nyxt will not start without ~/.local/share/nyxt/history/default.lisp file available.

**Information**
- OS name+version: Arch Linux
- Graphics card and driver: Nvidia
- Desktop environment / Window manager name+version: AwesomeWM Git version
- How you installed Nyxt (Guix pack, package manager, build from source): Build from source
- Information from `show-system-information`:
```
Nyxt version: 3-pre-release-2-191-gd6bc6d9dc
Renderer version: GI-GTK
Operating system kernel: Linux 6.0.2-arch1-1
Lisp implementation: SBCL 2.2.8 (Dynamic space size: 3221225472)
Features: (:NYXT-GI-GTK :NYXT-GTK :NYXT-GI-GTK :NYXT-GTK :WEBKIT2 :WEBKIT2-2.38 :WEBKIT2-CORS-ALLOWLIST :WEBKIT2-PASTE-PLAINTEXT :WEBKIT2-TRACKING :WEBKIT2-MUTE :WEBKIT2-EMOJI :WEBKIT2-MEDIA :WEBKIT2-SANDBOXING :GTK-3-22 :GTK-3-20 :GTK-3-18 :GTK-3-16 :GTK-3-14 :GTK-3-12 :GTK-3-10 :GTK-3-8 :GTK-3-6 :GTK-3-4 :GTK :GDK-3-22 :GDK-3-20 :GDK-3-18 :GDK-3-16 :GDK-3-14 :GDK-3-12 :GDK-3-10 :GDK-3-8 :GDK-3-6 :GDK-3-4 :CAIRO-1-10 :CAIRO-1-12 :GDK-PIXBUF :GLIB-2-30 :GLIB-2-32 :GLIB-2-34 :GLIB-2-36 :GLIB-2-38 :GLIB-2-40 :GLIB-2-42 :GLIB-2-44 :GLIB-2-46 :GLIB-2-48 :GLIB-2-50 :GLIB-2-52 :GLIB-2-54 :GLIB-2-56 :GLIB-2-58 :GLIB :NYXT-3.1 :NYXT-3 :NYXT-3-PRE-RELEASE-2 :NYXT-3-PRE-RELEASE-2-191-GD6BC6D9DC :CUSTOM-HASH-TABLE-NATIVE :SLYNK :PLUMP-UTF-32 :PARENSCRIPT :FSET-EXT-STRINGS :GLOBAL-VARS :DECLARE-TYPES :SBCL+SAFE-STANDARD-READTABLE :NAMED-READTABLES :SWANK :LPARALLEL :21BIT-CHARS :CLOSER-MOP :CL-PPCRE-UNICODE :CL-UNICODE :CL-JSON-DOUBLE-FLOAT-IS-SUBSUMED :CL-JSON-SINGLE-FLOAT-IS-SUBSUMED :CL-PPCRE :CHUNGA :FLEXI-STREAMS :BORDEAUX-THREADS :LPARALLEL.WITH-CLTL2 :LPARALLEL.WITH-CAS :LPARALLEL.WITH-STEALING-SCHEDULER :SPLIT-SEQUENCE CFFI-FEATURES:FLAT-NAMESPACE CFFI-FEATURES:X86-64 CFFI-FEATURES:UNIX :CFFI CFFI-SYS::FLAT-NAMESPACE ALEXANDRIA::SEQUENCE-EMPTYP :FAST-IO-SV :FAST-IO :CL-JSON-CLOS :CL-JSON :SBCL-USES-SB-ROTATE-BYTE CHIPZ-SYSTEM:GRAY-STREAMS :THREAD-SUPPORT :ASDF3.3 :ASDF3.2 :ASDF3.1 :ASDF3 :ASDF2 :ASDF :OS-UNIX :NON-BASE-CHARS-EXIST-P :ASDF-UNICODE :X86-64 :GENCGC :64-BIT :ANSI-CL :COMMON-LISP :ELF :IEEE-FLOATING-POINT :LINUX :LITTLE-ENDIAN :PACKAGE-LOCAL-NICKNAMES :SB-CORE-COMPRESSION :SB-LDB :SB-PACKAGE-LOCKS :SB-THREAD :SB-UNICODE :SBCL :UNIX)

ASDF version: 3.3.1
ASDF registries: (NYXT-SOURCE-REGISTRY ENVIRONMENT-SOURCE-REGISTRY)
Critical dependencies: (/home/build/nyxt-browser-git/src/nyxt-browser/_build/cl-cffi-gtk/gtk/cl-cffi-gtk.asd /home/build/nyxt-browser-git/src/nyxt-browser/_build/cl-gobject-introspection/cl-gobject-introspection.asd /home/build/nyxt-browser-git/src/nyxt-browser/_build/cl-webkit/webkit2/cl-webkit2.asd)
```

**Output when started from a shell**
```
(sbcl:8269): Gtk-WARNING **: 08:22:45.278: Theme parsing error: gtk.css:3611:22: 'none' is not a valid col
or name
Nyxt version 3-pre-release-2-191-gd6bc6d9dc
<INFO> [08:22:45] Source location: #P"/usr/local/share/nyxt/"
<INFO> [08:22:45] GTK extensions directory: #P"/usr/local/share/nyxt/libraries/web-extensions/"
<INFO> [08:22:45] Loading #P"/home/_user_/.local/share/nyxt/history/default.lisp".

```

Cheers and thanks for your awesome work!

### LINKED ISSUE #2624: Weird build issue producing binary that hangs early on startup
Here's a weird one.

Building current nyxt git on openSUSE Tumbleweed using SBCL 2.2.10 with **make all**, the produced nyxt binary 
hangs early on startup. See strace, it seems to mysteriously hang on a futex() call.
[nyxt_strace.txt](https://github.com/atlas-engineer/nyxt/files/9945807/nyxt_strace.txt)

After trying many things, including compiling with  previous SBCL versions (2.2.9 and 2.2.8, the latter I'm certain it worked at some point), this issue was still there.

When I was going to give up,  totally mystified about this issue, I accidentally found a workaround that fixes it:

```
make all
rm nyxt
make all
```

First **make all** compile the dependencies and build the nyxt binary.
The remaining 2 commands force only to rebuild the nyxt binary. For some reason it works, and the resulting 
binary launches fine.
 
So there's some weird issue causing a faulty binary when just **make all** is invoked to build deps + final binary.
It would be interesting if someone can reproduce this issue on his system.

### PR DESCRIPTION
**Pull request by Ambrevar**
# Description

This fixes a bunch of potential dead locks on startup, as well as some other bugs.

- virtual files no longer trigger the cache.
- `no-history-p` really disables the history of the buffer.
- `buffer-delete` no longer errors for buffers without history.
- Protect against history restoration errors on startup so that we ensure `*after-startup-hook*` and `startup-error-reporter-function` are run.

The potential dead lock is worked around by triggering the window creation first, then trigger the buffer creation + history restoration after waiting for the window creation to complete.

Fixes #2602 (hopefully), #2623 (should mitigate it).

Does _not_ fix the #2624 which seems to be linked to the build system.

Should help for  #2507, #2057.

# To test

You need the latest Nfiles.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [x] I have pulled from master before submitting this PR
- [x] There are no merge conflicts.
- [x] I've added the new dependencies as:
  - [x] ASDF dependencies,
  - [x] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [x] and Guix dependencies.
- [x] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [x] I have performed a self-review of my own code.
- [x] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [x] Documentation:
  - [x] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [x] I have updated the existing documentation to match my changes.
  - [x] I have commented my code in hard-to-understand areas.
  - [x] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [ ] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [x] Compilation and tests:
  - [x] My changes generate no new warnings.
  - [ ] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [x] New and existing unit tests pass locally with my changes.

### CODE REVIEW

File: source/browser.lisp [hunk 1]
```diff
@@ -266,52 +276,62 @@ prevents otherwise.")
...
-    (mapcar #'window-delete (window-list))
-    (window-make browser)
     (if (restore-session-on-startup-p *browser*)
-        (if (restore-history-buffers (files:content (history-file *browser*))
-                                     (history-file *browser*))
+        (if (with-protect ("Error restoring history ~a: ~a"
+                           (files:expand (history-file *browser*))
+                           :condition)
+              (restore-history-buffers
+               (files:content (history-file *browser*))
```
Comment (aartaka) [inline source/browser.lisp:333]: But then, if error happens in the `files:content` call, then the case of `restore-session-on-startup-p` set to NIL will not be covered, right?

File: source/buffer.lisp [hunk 2]
```diff
@@ -1089,11 +1093,12 @@ See `make-buffer' for a description of the arguments."
 
 (defmethod buffer-delete ((buffer context-buffer))
   (files:with-file-content (history (history-file buffer))
-    (sera:and-let* ((owner (htree:owner history (id buffer)))
-                    (current (htree:current owner))
-                    (data (htree:data current)))
-      (setf (nyxt::scroll-position data) (nyxt:document-scroll-position buffer))
-      (htree:delete-owner history (id buffer))))
+    (when history
+      (sera:and-let* ((owner (htree:owner history (id buffer)))
```
Comment (aartaka) [inline source/buffer.lisp:1097]: This position rememberance does not work, and we probably should remove it...

**Review by aartaka — COMMENTED** [review]
It looks slightly weird, but it does indeed fix the startup errors for history.

File: source/buffer.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/buffer.lisp:1097]: I remember you removed it some time ago, how come it's still there?  Or am I confused with something else?

**Comment by Ambrevar** [comment]
> It looks slightly weird, but it does indeed fix the startup errors for history.

What do you mean, "weird"?

File: source/buffer.lisp (hunk 2 above)
Comment (aartaka) [reply source/buffer.lisp:1097]: Yes, I remember removing it too. Some rebase mistake maybe?

**Comment by aartaka** [comment]
> What do you mean, "weird"?

I mean, this PR introduces some two more layers (`finalize-window`, `finalize-history`) to our already complex startup sequence. Which feels wrong. But I cannot see any better way to avoid history breaking the browser, so I guess that's a reasonable sacrifice for the sake of stability.

File: source/browser.lisp (hunk 1 above)
Comment (Ambrevar) [reply source/browser.lisp]: Done in e68f49c30d1574a7b1a0828ad6fc1693685362f0.

**Comment by Ambrevar** [comment]
If you look at it like sequential startup steps, then it does not change anything except ensure that the steps are run at the right time with respect to the readiness of the renderer.

**Comment by Ambrevar** [comment]
(Also I find that calling everything `finalize*` helps understand what's going on.  Previously the `startup` name was really confusing.)

**Comment by aartaka** [comment]
> If you look at it like sequential startup steps, then it does not change anything except ensure that the steps are run at the right time with respect to the readiness of the renderer.

Fair!

**Comment by Ambrevar** [comment]
OK to merge?  @aadcg any comment?

**Comment by aartaka** [comment]
OK to merge on my side :)

**Comment by aartaka** [comment]
@Ambrevar, don't forget to update the changelog mentioning the improved history stability!

**Comment by aartaka** [comment]
And go through the checklist to make sure everything's covered—it's there to help :P

**Comment by Ambrevar** [comment]
> And go through the checklist to make sure everything's covered—it's there to help :P

Already done, what do you mean?

**Comment by Ambrevar** [comment]
There is no change in API, why mention it in the changelog?

**Comment by aartaka** [comment]
Because it's an improvement over how the corrupted history was influencing the startup experience before!

**Comment by Ambrevar** [comment]
Fair enough, done :)

**Comment by aartaka** [comment]
THen yes, I have nothing to complain about anymore :)

**Comment by Ambrevar** [comment]
Merged with 882ce0786b79052c5e9dcf45a31079e70d92f324.

**Comment by aartaka** [comment]
Cool, thanks!

**Pull request closed without merging**